File: track/core/time_entry.py
import time
from dataclasses import dataclass

from track.core.helpers import TimeHelper


@dataclass
class TimeEntry:
    id: int
    pid: int
    wid: int
    description: str
    _duration: int

    @property
    def duration(self) -> int:
        if self._duration < 0:
            return int(time.time()) + self._duration
        return self._duration

    @duration.setter
    def duration(self, value: int) -> None:
        self._duration = value

    def __str__(self):
        return f"{TimeHelper.seconds_to_hms_string(self.duration)} - {self.description}"

File: test/builders/time-entry.ts
import {Duration, TimeEntry, TimeEntryParams} from "../../src/core";

type TimeEntryBuilderParams = Partial<Omit<TimeEntryParams, 'duration'> & {
    duration: number
}>
export function buildTimeEntry({description, duration, id, pid, wid}: TimeEntryBuilderParams) {
    return new TimeEntry({
        description: description ?? 'Test entry',
        duration: new Duration(duration ?? 1800),
        id: id ?? Math.floor(Math.random() * 10 ** 16),
        pid: pid ?? 1,
        wid: wid ?? 1
    })
}
